core/corehttp/gateway_handler.go

```diff
@@ -133,6 +133,7 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
 	}
 
 	w.Header().Set("X-IPFS-Path", urlPath)
+	w.Header().Set("Access-Control-Allow-Origin", "*")
```

Comment:
> - `Access-Control-Allow-Credentials` is just for cookies & HTTP-Basic auth, not a more modern / explicit method of authenticating like oAuth.

indeed!

> - why would we likely have authentication on a gateway? wouldn't encryption of the resources themselves be the preferred method of keeping data safe on a p2p network?

yes, but some people may want to enable simple auth. we shouldn't prevent them.

All i'm saying is that instead of adding a forced `*` we just allow the user to specify the headers via a config like:

```
ipfs config Gateway.CORS "*"
# or 
ipfs config Gateway.CORS "<origin>"
# or
ipfs config Gateway.CORS "mysite.com"
```
or maybe

```
ipfs config Gateway.Headers.Access-Control-Allow-Origin "*"
# or 
ipfs config Gateway.Headers.Access-Control-Allow-Origin "<origin>"
# or
ipfs config Gateway.Headers.Access-Control-Allow-Origin "mysite.com"
```

And if we do this, we have a much more general solution, that allows all sorts of other headers to be set. this is the more useful way to implement this.

> - gateway operators wanting to restrict the origins that they allow seems kinda silly, but they can still totally do that with a reverse-proxy like nginx in front of ipfs... I'd be happy to write up a guide if this is something that people actually want to do.

nginx and proxies are great, and certainly can be used this way. but we shouldn't _force_ this to be the case. enabling CORS can allow lots of cross-site scripting problems.

> - how would this apply to non-browsers? when you use cURL it doesn't hide the response from you when `null` isn't a whitelisted origin... CORS is enforced by the browser and only in the browser.

We should endeavor to keep things consistent. there's all sorts of tools that need to be able to get the same headers. proxies, tunnels, headless browsers, pipes, ... i've definitely been annoyed in the past when a `curl` that i piped into an experimental browser-renderer didnt give me all the same headers it have chrome.